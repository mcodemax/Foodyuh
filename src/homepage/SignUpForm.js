import React, { useState, useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import './SignUpForm.scss';
import { Formik, Form, Field, ErrorMessage } from 'formik';
import * as Yup from 'yup';
import { v4 as uuidv4 } from 'uuid';

/** generates signup form
 * I: signUp function to change state of logged in or not
 * { username, password, firstName, lastName, email }
 */
const SignUpForm = ({ signUp }) => {
  const [signUpErrors, setSignUpErrors] = useState([]);
  const navigate = useNavigate();
  const initialValues = {
    username: '',
    firstName: '',
    lastName: '',
    email: '',
    password: '',
    confirmPassword: '',
  };

  const validationSchema = Yup.object({
    username: Yup.string()
      .min(1)
      .max(25, 'Username cannot exceed 25 characters')
      .required('Required'),
    firstName: Yup.string()
      .min(1)
      .max(255, 'Max first name length is 255 characters')
      .required('Required'),
    lastName: Yup.string()
      .min(1)
      .max(255, 'Max last name length is 255 characters')
      .required('Required'),
    email: Yup.string()
      .email('Must be a valid email')
      .min(1)
      .max(255, 'Max email length is 255 characters')
      .required('Email is required'),
    password: Yup.string()
      .min(5)
      .max(255, 'Max password length is 255 characters')
      .required('Required'),
    confirmPassword: Yup.string().when('password', {
      is: (val) => (val && val.length > 0 ? true : false),
      then: Yup.string().oneOf(
        [Yup.ref('password')],
        'Both passwords need to match'
      ),
    }),
  });

  const onSubmit = async (values, { resetForm }) => {
    console.debug('In SignUpForm.js, onSubmit');
    console.log('Form Data: ', values); //comment out later so password not revealed

    //a copy of values object is made so this component does not become uncontrolled
    const valuesCopy = { ...values };
    delete valuesCopy.confirmPassword;

    const res = await signUp(valuesCopy);

    if (res.success) {
      //maybe make success banner if true in html
      navigate('/user/profile');

      //see bug in todos
    } else {
      setSignUpErrors(res.errors); //trouble shoot later how to actuallyy display these properly
      resetForm({});
    }
    //if login fail show error message and navigate back2Login

    // https://github.com/jaredpalmer/formik/issues/446
  };

  return (
    <>
      <div className='SignUpForm'>
        {signUpErrors && signUpErrors.length
          ? signUpErrors.map((error) => {
              return (
                <div className='SignUpForm-failed' key={uuidv4()}>
                  {error}
                </div>
              );
            })
          : ``}
        <Formik
          initialValues={initialValues}
          validationSchema={validationSchema}
          onSubmit={onSubmit}
        >
          <>
            {/* Formik component can only have one child. */}
            <p>SIGN UP!</p>
            <Form className='SignUpForm-form'>
              <div className='SignUpForm-username field'>
                <label htmlFor='username'>Username</label>
                <Field type='text' id='username' name='username' />
                <p className='form-warning'><ErrorMessage name='username' /></p>
              </div>
              <div className={`SignUpForm-firstName field`}>
                <label htmlFor='firstName'>First Name</label>
                <Field type='text' id='firstName' name='firstName' />
                <p className='form-warning'><ErrorMessage name='firstName'/></p>
              </div>
              <div className={`SignUpForm-lastName field`}>
                <label htmlFor='lastName'>Last Name</label>
                <Field type='text' id='lastName' name='lastName' />
                <p className='form-warning'><ErrorMessage name='lastName' /></p>
              </div>
              <div className={`SignUpForm-email field`}>
                <label htmlFor='email'>Email</label>
                <Field type='text' id='email' name='email' />
                <p className='form-warning'><ErrorMessage name='email' /></p>
              </div>
              <div className={`SignUpForm-password field`}>
                <label htmlFor='password'>Password</label>
                <Field type='password' id='password' name='password' />
                <p className='form-warning'><ErrorMessage name='password' autoComplete='off' /></p>
              </div>
              <div className={`SignUpForm-confirmPassword field`}>
                <label htmlFor='confirmPassword'>Confirm Password</label>
                <Field
                  type='password'
                  id='confirmPassword'
                  name='confirmPassword'
                />
                <p className='form-warning'><ErrorMessage name='confirmPassword' autoComplete='off' /></p>
              </div>
              <button type='submit' id='submit'>
                Submit
              </button>
            </Form>
          </>
        </Formik>
      </div>
    </>
  );
};

export default SignUpForm;
