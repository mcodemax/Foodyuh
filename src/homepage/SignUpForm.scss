.SignUpForm {
    font-size: 1rem;
    font-weight: 400;
    line-height: 1.5;
    color: #c7d6e6;
    font-family: Verdana, Geneva, Tahoma, sans-serif;
    -webkit-font-smoothing: antialiased;
    box-sizing: border-box;
    width: auto;
    padding-right: 15px;
    padding-left: 15px;
    margin-right: auto;
    margin-left: auto;
    max-width: 720px;
    text-align: center!important;

    .SignUpForm-form {
        display: flex;
        flex-direction: column;
        box-sizing: border-box;
        
        label {
            font-family: Verdana, Geneva, Tahoma, sans-serif;
            text-align: left;
        }

        input {
            border-radius: .25rem;
            font-family: Geneva, sans-serif;
        }

        .field {
            font-family: Geneva, sans-serif;
            display: flex;
            flex-direction: column;
            height: 90px;
        }

        .form-warning {
            font-family: Geneva, sans-serif;
            color: red;
            margin-top: 0;
        }
    }
    

    // label {
    //     display: inline-block;
    //     text-align: center;
    //     vertical-align: middle;
    //     word-wrap: break-word;
    //     color: inherit;
    //     box-sizing: border-box;
    //     margin-top: 20px;
    //     font-weight: 500;
    //     line-height: 1.2;
    //     font-size: 1rem;
    //     margin-bottom: .75rem;
    // }

    // input {
    //     color: #495057;
    //     display: inline-block;
    //     background-color: #fff;
    //     background-clip: padding-box;
    //     border: 1px solid #ced4da;
    //     border-radius: 10rem;
    //     width: auto;
    //     padding: .25rem .6rem;
    //     vertical-align: middle;
    // }   

    button {
        // box-sizing: border-box;
        text-decoration: none;
        display: inline-block;
        text-align: center;
        vertical-align: middle;
        user-select: none;
        // border: 1px solid transparent;
        padding: .3rem .75rem;
        font-size: 1rem;
        line-height: 1.5;
        border-radius: .25rem;
        color: #fff;
        background-color: #007bff;
        border-color: #007bff;
        margin-top: 1rem!important;
        font-weight: 600!important;
        cursor: pointer;       
    }
}