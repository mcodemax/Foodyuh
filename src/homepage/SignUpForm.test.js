import React from "react";
import { render } from "@testing-library/react";
import SignUpForm from "./SignUpForm";

const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
   useNavigate: () => mockedUsedNavigate,
}));

it("renders without crashing", function () {
  render(<SignUpForm />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<SignUpForm />);
  expect(asFragment()).toMatchSnapshot();
});
