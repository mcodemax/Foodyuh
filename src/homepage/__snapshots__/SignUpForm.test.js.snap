// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="SignUpForm"
  >
    <p>
      SIGN UP!
    </p>
    <form
      action="#"
      class="SignUpForm-form"
    >
      <div
        class="SignUpForm-username field"
      >
        <label
          for="username"
        >
          Username
        </label>
        <input
          id="username"
          name="username"
          type="text"
          value=""
        />
        <p
          class="form-warning"
        />
      </div>
      <div
        class="SignUpForm-firstName field"
      >
        <label
          for="firstName"
        >
          First Name
        </label>
        <input
          id="firstName"
          name="firstName"
          type="text"
          value=""
        />
        <p
          class="form-warning"
        />
      </div>
      <div
        class="SignUpForm-lastName field"
      >
        <label
          for="lastName"
        >
          Last Name
        </label>
        <input
          id="lastName"
          name="lastName"
          type="text"
          value=""
        />
        <p
          class="form-warning"
        />
      </div>
      <div
        class="SignUpForm-email field"
      >
        <label
          for="email"
        >
          Email
        </label>
        <input
          id="email"
          name="email"
          type="text"
          value=""
        />
        <p
          class="form-warning"
        />
      </div>
      <div
        class="SignUpForm-password field"
      >
        <label
          for="password"
        >
          Password
        </label>
        <input
          id="password"
          name="password"
          type="password"
          value=""
        />
        <p
          class="form-warning"
        />
      </div>
      <div
        class="SignUpForm-confirmPassword field"
      >
        <label
          for="confirmPassword"
        >
          Confirm Password
        </label>
        <input
          id="confirmPassword"
          name="confirmPassword"
          type="password"
          value=""
        />
        <p
          class="form-warning"
        />
      </div>
      <button
        id="submit"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
